### PR TITLE
Organizations

### PR DESCRIPTION
**Pull request by bonifaido**
This PR will introduce the basic organisation handling of Pipeline. It includes:
- New API endpoints with orgs/:orgId:
- Default organisation for each user (Github name)
- Secret handling (AWS, AKS, GKE) with organisation scope

### CODE REVIEW

File: api/cluster.go [hunk 1]
```diff
@@ -420,7 +423,9 @@ func FetchClusters(c *gin.Context) {
 
 	var clusters []model.ClusterModel //TODO change this to CommonClusterStatus
 	db := model.GetDB()
-	err := db.Find(&clusters).Error
+	organization := auth.GetCurrentOrganization(c.Request)
+	organization.Name = ""
```
Comment (baluchicken) [inline api/cluster.go]: Is this needed here? By default string will be initialized as ""

File: api/organization.go [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+	"github.com/sirupsen/logrus"
+)
+
+//OrganizationMiddleware parses the organization id from the request, queries it from the database and saves it to the current context
+func OrganizationMiddleware(c *gin.Context) {
+	orgidParam := c.Param("orgid")
+	orgid, err := strconv.ParseUint(orgidParam, 10, 32)
+	if err != nil {
+		message := "Error parsing organization id"
+		log.Info(message, ": ", err)
```
Comment (baluchicken) [inline api/organization.go]: Is this really an Info? shouldn't we use err.Error()?

File: api/organization.go [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+		c.AbortWithStatusJSON(http.StatusBadRequest, components.ErrorResponse{
+			Code:    http.StatusBadRequest,
+			Message: message,
+			Error:   message,
```
Comment (baluchicken) [inline api/organization.go]: is this the really the message? not the err?

File: api/organization.go [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+
+	user := auth.GetCurrentUser(c.Request)
+	var organization = auth.Organization{ID: uint(orgid)}
+	var organizations []auth.Organization
+
+	db := model.GetDB()
+	err = db.Model(user).Where(&organization).Related(&organizations, "Organizations").Error
+	if err != nil {
+		message := "Error listing organizations"
+		log.Info(message, ": ", err)
```
Comment (baluchicken) [inline api/organization.go]: Is this really an Info? shouldn't we use err.Error()?

File: api/organization.go [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+		c.AbortWithStatusJSON(http.StatusInternalServerError, components.ErrorResponse{
+			Code:    http.StatusInternalServerError,
+			Message: message,
+			Error:   message,
```
Comment (baluchicken) [inline api/organization.go]: is this the really the message? not the err?

File: api/organization.go [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+		})
+		return
+	}
+
+	if len(organizations) != 1 {
+		message := "Organization not found"
+		log.Info(message)
```
Comment (baluchicken) [inline api/organization.go]: Is this really info?

File: cluster/gke.go [hunk 7]
```diff
@@ -356,31 +354,50 @@ func (g *GKECluster) GetModel() *model.ClusterModel {
...
-}
-
-func getGoogleServiceClient() (*gke.Service, error) {
+	// Get Secret from Vault
+	clusterSecret, err := GetSecret(g)
+	if err != nil {
+		return nil, err
+	}
+	if clusterSecret.SecretType != secret.Google {
+		return nil, errors.Errorf("missmatch secret type %s versus %s", clusterSecret.SecretType, secret.Amazon)
```
Comment (tarokkk) [inline cluster/gke.go]: typo: secret.Amazon -> secret.Google

File: api/organization.go [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+
+	"github.com/banzaicloud/banzai-types/components"
+	"github.com/banzaicloud/pipeline/auth"
+	"github.com/banzaicloud/pipeline/model"
+	"github.com/gin-gonic/gin"
+	"github.com/sirupsen/logrus"
+)
+
+//OrganizationMiddleware parses the organization id from the request, queries it from the database and saves it to the current context
+func OrganizationMiddleware(c *gin.Context) {
```
Comment (baluchicken) [inline api/organization.go]: Initialization of the logger is missing.

File: api/organization.go [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+	log := logger.WithFields(logrus.Fields{"tag": "GetOrganizations"})
+	log.Info("Fetching organizations")
+
+	user := auth.GetCurrentUser(c.Request)
+
+	idParam := c.Param("id")
+	id, err := strconv.ParseUint(idParam, 10, 32)
+	if idParam != "" && err != nil {
+		message := "Error parsing organization id"
+		log.Info(message, ": ", err)
```
Comment (baluchicken) [inline api/organization.go]: Is this really an Info? shouldn't we use err.Error()?

File: api/organization.go [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+		c.JSON(http.StatusBadRequest, components.ErrorResponse{
+			Code:    http.StatusBadRequest,
+			Message: message,
+			Error:   message,
```
Comment (baluchicken) [inline api/organization.go]: is this the really the message? not the err?

File: api/organization.go [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+		return
+	}
+
+	var organization = auth.Organization{ID: uint(id)}
+	var organizations []auth.Organization
+	db := model.GetDB()
+	err = db.Model(user).Where(&organization).Related(&organizations, "Organizations").Error
+	if err != nil {
+		message := "Error listing organizations"
+		log.Info(message, ": ", err)
```
Comment (baluchicken) [inline api/organization.go]: Is this really an Info? shouldn't we use err.Error()?

File: api/organization.go [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+		c.JSON(http.StatusInternalServerError, components.ErrorResponse{
+			Code:    http.StatusInternalServerError,
+			Message: message,
+			Error:   message,
```
Comment (baluchicken) [inline api/organization.go]: is this the really the message? not the err?

File: api/organization.go [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+
+//CreateOrganization creates an organization for the calling user
+func CreateOrganization(c *gin.Context) {
+	log := logger.WithFields(logrus.Fields{"tag": "CreateOrganization"})
+	log.Info("Creating organization")
+
+	user, err := auth.GetCurrentUserFromDB(c.Request)
+	if err != nil {
+		message := "Error creating organization"
+		log.Info(message, ": ", err)
```
Comment (baluchicken) [inline api/organization.go]: Is this really an Info? shouldn't we use err.Error()?

File: api/organization.go [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+		c.JSON(http.StatusInternalServerError, components.ErrorResponse{
+			Code:    http.StatusInternalServerError,
+			Message: message,
+			Error:   message,
```
Comment (baluchicken) [inline api/organization.go]: is this the really the message? not the err?

File: api/organization.go [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+		})
+		return
+	}
+
+	var name struct {
+		Name string
+	}
+
+	if err := c.ShouldBindJSON(&name); err != nil {
```
Comment (baluchicken) [inline api/organization.go]: this can be simplified c.ShouldBindJson(&name) != nil

File: api/organization.go [hunk 16]
```diff
@@ -0,0 +1,148 @@
...
+		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
+		return
+	}
+
+	organization := auth.Organization{Name: name.Name, Users: []auth.User{*user}}
+	db := model.GetDB()
+	err = db.Save(&organization).Error
+	if err != nil {
+		message := "Error creating organization"
+		log.Info(message, ": ", err)
```
Comment (baluchicken) [inline api/organization.go]: Is this really an Info? shouldn't we use err.Error()?

File: api/organization.go [hunk 17]
```diff
@@ -0,0 +1,148 @@
...
+		c.JSON(http.StatusBadRequest, components.ErrorResponse{
+			Code:    http.StatusBadRequest,
+			Message: message,
+			Error:   message,
```
Comment (baluchicken) [inline api/organization.go]: is this the really the message? not the err?

File: api/secrets.go [hunk 18]
```diff
@@ -0,0 +1,121 @@
...
+	"net/http"
+
+	"github.com/banzaicloud/banzai-types/components"
+	"github.com/banzaicloud/pipeline/auth"
+	"github.com/banzaicloud/pipeline/secret"
+	"github.com/gin-gonic/gin"
+	"github.com/sirupsen/logrus"
+)
+
+func AddSecrets(c *gin.Context) {
```
Comment (baluchicken) [inline api/secrets.go]: Doc is missing:) please keep golint happy:)

File: api/secrets.go [hunk 19]
```diff
@@ -0,0 +1,121 @@
...
+
+	log.Info("Get organization id from params")
+	organizationID := auth.GetCurrentOrganization(c.Request).IDString()
+	log.Infof("Organization id: %s", organizationID)
+
+	log.Info("Binding request")
+
+	var createSecretRequest secret.CreateSecretRequest
+	if err := c.ShouldBind(&createSecretRequest); err != nil {
+		log.Errorf("Error during binding CreateSecretRequest: %s", err)
```
Comment (baluchicken) [inline api/secrets.go]: use err.Error()

File: api/secrets.go [hunk 20]
```diff
@@ -0,0 +1,121 @@
...
+	log.Infof("Secret stored at: %s", secretPath)
+
+	c.JSON(http.StatusCreated, secret.CreateSecretResponse{
+		Name:       createSecretRequest.Name,
+		SecretType: createSecretRequest.SecretType,
+		SecretID:   secretID,
+	})
+}
+
+func ListSecrets(c *gin.Context) {
```
Comment (baluchicken) [inline api/secrets.go]: Doc is missing:) please keep golint happy.

File: api/secrets.go [hunk 21]
```diff
@@ -0,0 +1,121 @@
...
+		})
+	} else {
+		log.Infof("Listing secrets succeeded: %#v", items)
+		c.JSON(http.StatusOK, secret.ListSecretsResponse{
+			Secrets: items,
+		})
+	}
+}
+
+func DeleteSecrets(c *gin.Context) {
```
Comment (baluchicken) [inline api/secrets.go]: Doc is missing

File: api/secrets.go [hunk 22]
```diff
@@ -0,0 +1,121 @@
...
+	log = logger.WithFields(logrus.Fields{"tag": "Delete Secrets"})
+	log.Info("Start deleting secrets")
+
+	log.Info("Get organization id from params")
```
Comment (baluchicken) [inline api/secrets.go]: This three line should be factored out to a helper function. I saw this the third time.

File: api/cluster.go (hunk 1 above)
Comment (tarokkk) [reply api/cluster.go]: This is not a simple object, it's parsed from the request. For further query the name is ignored only the Id used

File: auth/user.go [hunk 23]
```diff
@@ -40,18 +48,55 @@ type DroneUser struct {
...
+	ID        uint                 `gorm:"primary_key" json:"id"`
+	CreatedAt time.Time            `json:"createdAt"`
+	UpdatedAt time.Time            `json:"updatedAt"`
+	DeletedAt *time.Time           `sql:"index" json:"deletedAt,omitempty"`
+	Name      string               `gorm:"unique;not null" json:"name"`
+	Users     []User               `gorm:"many2many:user_organizations" json:"users,omitempty"`
+	Clusters  []model.ClusterModel `gorm:"foreignkey:organization_id" json:"clusters,omitempty"`
+}
+
+func (org *Organization) IDString() string {
```
Comment (baluchicken) [inline auth/user.go]: Doc is missing

File: auth/user.go [hunk 24]
```diff
@@ -40,18 +48,55 @@ type DroneUser struct {
...
+	return strconv.FormatUint(uint64(org.ID), 10)
+}
+
 //TableName sets DroneUser's table name
 func (DroneUser) TableName() string {
 	return "users"
 }
 
-func getCurrentUser(req *http.Request) *User {
+func GetCurrentUser(req *http.Request) *User {
```
Comment (baluchicken) [inline auth/user.go]: Doc is missing

File: auth/user.go [hunk 25]
```diff
@@ -40,18 +48,55 @@ type DroneUser struct {
...
 	if currentUser, ok := Auth.GetCurrentUser(req).(*User); ok {
 		return currentUser
 	}
 	return nil
 }
 
+func GetCurrentOrganization(req *http.Request) *Organization {
```
Comment (baluchicken) [inline auth/user.go]: Doc is missing

File: auth/user.go [hunk 26]
```diff
@@ -40,18 +48,55 @@ type DroneUser struct {
...
+	if organization := req.Context().Value(CurrentOrganization); organization != nil {
+		return organization.(*Organization)
+	}
+	return nil
+}
+
+func GetCurrentUserFromDB(req *http.Request) (*User, error) {
```
Comment (baluchicken) [inline auth/user.go]: Doc is missing

File: auth/user.go [hunk 27]
```diff
@@ -40,18 +48,55 @@ type DroneUser struct {
...
+	if currentUser, ok := Auth.GetCurrentUser(req).(*User); ok {
+		claims := &claims.Claims{UserID: strconv.Itoa(int(currentUser.ID))}
+		context := &auth.Context{Auth: Auth, Claims: claims, Request: req}
+		if user, err := Auth.UserStorer.Get(claims, context); err == nil {
+			return user.(*User), nil
+		}
+	}
+	return nil, nil
```
Comment (baluchicken) [inline auth/user.go]: shouldn't we return the error here?

File: auth/user.go [hunk 28]
```diff
@@ -40,18 +48,55 @@ type DroneUser struct {
 	Synced int64  `gorm:"column:user_synced"`
 }
 
+type UserOrganization struct {
```
Comment (baluchicken) [inline auth/user.go]: Doc is missing

File: cluster/aks.go [hunk 29]
```diff
@@ -44,6 +47,35 @@ type AKSCluster struct {
 	APIEndpoint  string
 }
 
+func (c *AKSCluster) GetOrg() uint {
```
Comment (baluchicken) [inline cluster/aks.go]: Doc is missing

File: cluster/aks.go [hunk 30]
```diff
@@ -44,6 +47,35 @@ type AKSCluster struct {
...
+	return c.modelCluster.OrganizationId
+}
+
+func (c *AKSCluster) GetAKSClient() (*azureClient.AKSClient, error) {
```
Comment (baluchicken) [inline cluster/aks.go]: Doc is missing

File: cluster/aks.go [hunk 31]
```diff
@@ -44,6 +47,35 @@ type AKSCluster struct {
...
+	if err != nil {
+		return nil, err
+	}
+	if clusterSecret.SecretType != secret.Azure {
+		return nil, errors.Errorf("missmatch secret type %s versus %s", clusterSecret.SecretType, secret.Azure)
+	}
+	creds := &azureCluster.AKSCredential{
+		ClientId:       clusterSecret.Values["AZURE_CLIENT_ID"],
+		ClientSecret:   clusterSecret.Values["AZURE_CLIENT_SECRET"],
+		SubscriptionId: clusterSecret.Values["AZURE_TENANT_ID"],
```
Comment (baluchicken) [inline cluster/aks.go]: It is in a wrong order:) the tenant id is switched with the subscription id

File: cluster/aks.go [hunk 32]
```diff
@@ -44,6 +47,35 @@ type AKSCluster struct {
...
+		TenantId:       clusterSecret.Values["AZURE_SUBSCRIPTION_ID"],
+	}
+	client, err := azureClient.GetAKSClient(creds)
+	if err != nil {
+		return nil, err
+	}
+	return client, nil
+}
+
+func (c *AKSCluster) GetSecretID() string {
```
Comment (baluchicken) [inline cluster/aks.go]: Doc is missing

File: cluster/aws.go [hunk 33]
```diff
@@ -34,8 +37,11 @@ func init() {
 	log = logger.WithFields(logrus.Fields{"action": "Cluster"})
 }
 
-var runtimeParam = cutil.RuntimeParameters{
-	AwsProfile: "",
+func SetCredentials(awscred *credentials.Credentials) func(*session.Options) error {
```
Comment (baluchicken) [inline cluster/aws.go]: Doc is missing

File: cluster/aws.go [hunk 34]
```diff
@@ -46,6 +52,14 @@ type AWSCluster struct {
 	APIEndpoint     string
 }
 
+func (c *AWSCluster) GetOrg() uint {
```
Comment (baluchicken) [inline cluster/aws.go]: Doc is missing

File: cluster/aws.go [hunk 35]
```diff
@@ -46,6 +52,14 @@ type AWSCluster struct {
...
+	return c.modelCluster.OrganizationId
+}
+
+func (c *AWSCluster) GetSecretID() string {
```
Comment (baluchicken) [inline cluster/aws.go]: Doc is missing

File: cluster/common.go [hunk 36]
```diff
@@ -28,14 +30,22 @@ type CommonCluster interface {
...
 	AddDefaultsToUpdate(*bTypes.UpdateClusterRequest)
 	GetAPIEndpoint() (string, error)
 	DeleteFromDatabase() error
-	//ModifyCluster(*model.ClusterModel)
-	//GetKubernetesConf()
-	//GetKubernetesEndpoint()
+	GetOrg() uint
+}
+
+func GetSecret(cluster CommonCluster) (*secret.SecretsItemResponse, error) {
```
Comment (baluchicken) [inline cluster/common.go]: Doc is missing

File: cluster/common.go [hunk 37]
```diff
@@ -28,14 +30,22 @@ type CommonCluster interface {
...
+	org := strconv.FormatUint(uint64(cluster.GetOrg()), 10)
+	secret, err := secret.Store.Get(org, cluster.GetSecretID())
+	if err != nil {
```
Comment (baluchicken) [inline cluster/common.go]: this can be simplified, by just returning the secret.Store.Get(org, cluster.GetSecretID())

File: cluster/gke.go [hunk 38]
```diff
@@ -43,8 +41,21 @@ const (
 	netesDefault     = "netes-default"
 )
 
+type ServiceAccount struct {
```
Comment (baluchicken) [inline cluster/gke.go]: Doc is missing

File: cluster/gke.go [hunk 39]
```diff
@@ -73,12 +86,19 @@ type GKECluster struct {
 	APIEndpoint   string
 }
 
+func (g *GKECluster) GetOrg() uint {
```
Comment (baluchicken) [inline cluster/gke.go]: Doc is missing

File: cluster/gke.go [hunk 40]
```diff
@@ -73,12 +86,19 @@ type GKECluster struct {
...
+	return g.modelCluster.OrganizationId
+}
+
+func (g *GKECluster) GetSecretID() string {
```
Comment (baluchicken) [inline cluster/gke.go]: Doc is missing

File: api/organization.go [hunk 41]
```diff
@@ -0,0 +1,150 @@
...
+		})
+		return
+	}
+
+	var name struct {
+		Name string
+	}
+
+	if err := c.ShouldBindJSON(&name); err != nil {
+		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
```
Comment (colin014) [inline api/organization.go]: this should be an ErrorResponse

File: cluster/gke.go [hunk 42]
```diff
@@ -1073,3 +1084,64 @@ func (g *GKECluster) DeleteFromDatabase() error {
...
+		} else {
+			log.Info("Getting server config succeeded")
+			c.JSON(http.StatusOK, convertServerConfig(serverConfig))
+		}
+
+	}
+
+}
+
+type GetServerConfigResponse struct {
```
Comment (baluchicken) [inline cluster/gke.go]: Doc is missing

File: cluster/hooks.go [hunk 43]
```diff
@@ -70,9 +70,9 @@ func PersistKubernetesKeys(cluster CommonCluster) {
 	configMapName := viper.GetString("monitor.configmap")
 	configMapPath := viper.GetString("monitor.mountPath")
 	if configMapName != "" && configMapPath != "" {
-		log.Infof("Save certificates to configmap: %s", configMapName)
+		log.Infof("save certificates to configmap: %s", configMapName)
 		if err := saveKeysToConfigmap(config, configMapName, cluster.GetName()); err != nil {
-			log.Errorf("Error saving certs to configmap: ", err)
+			log.Errorf("error saving certs to configmap: %s", err)
```
Comment (baluchicken) [inline cluster/hooks.go]: err.Error()

File: secret/store.go [hunk 44]
```diff
@@ -0,0 +1,182 @@
...
+	General    = "GENERAL_SECRET"
+	Kubernetes = "KUBERNETES_SECRET"
+)
+
+func init() {
+	logger = config.Logger()
+	Store = newVaultSecretStore()
+}
+
+type SecretStore struct {
```
Comment (baluchicken) [inline secret/store.go]: Doc is missing

File: secret/store.go [hunk 45]
```diff
@@ -0,0 +1,182 @@
...
+	client  *vault.Client
+	logical *vaultapi.Logical
+}
+
+type CreateSecretResponse struct {
```
Comment (baluchicken) [inline secret/store.go]: Doc is missing

File: secret/store.go [hunk 46]
```diff
@@ -0,0 +1,182 @@
...
+	Name       string `json:"name" binding:"required"`
+	SecretType string `json:"type" binding:"required"`
+	SecretID   string `json:"secret_id"`
+}
+
+type CreateSecretRequest struct {
```
Comment (baluchicken) [inline secret/store.go]: Doc is missing

File: secret/store.go [hunk 47]
```diff
@@ -0,0 +1,182 @@
...
+	Name       string            `json:"name" binding:"required"`
+	SecretType string            `json:"type" binding:"required"`
+	Values     map[string]string `json:"values" binding:"required"`
+}
+
+type ListSecretsResponse struct {
```
Comment (baluchicken) [inline secret/store.go]: Doc is missing

File: secret/store.go [hunk 48]
```diff
@@ -0,0 +1,182 @@
...
+	Secrets []SecretsItemResponse `json:"secrets"`
+}
+
+type SecretsItemResponse struct {
```
Comment (baluchicken) [inline secret/store.go]: Doc is missing

File: secret/store.go [hunk 49]
```diff
@@ -0,0 +1,182 @@
...
+	role := "pipeline"
+	client, err := vault.NewClient(role)
+	if err != nil {
+		panic(err)
+	}
+	logical := client.Vault().Logical()
+	return &SecretStore{client: client, logical: logical}
+}
+
+func GenerateSecretID() string {
```
Comment (baluchicken) [inline secret/store.go]: Doc is missing

File: secret/store.go [hunk 50]
```diff
@@ -0,0 +1,182 @@
...
+		"client_email",
+		"client_id",
+		"auth_uri",
+		"token_uri",
+		"auth_provider_x509_cert_url",
+		"client_x509_cert_url",
+	},
+}
+
+func (c *CreateSecretRequest) Validate() error {
```
Comment (baluchicken) [inline secret/store.go]: Doc is missing

File: secret/store.go [hunk 51]
```diff
@@ -0,0 +1,182 @@
...
+	for _, key := range requiresKeys {
+		if _, ok := c.Values[key]; !ok {
+			return errors.Errorf("missing key: %s", key)
+		}
+
+	}
+	return nil
+}
+
+func (ss *SecretStore) Delete(organizationID, secretID string) error {
```
Comment (baluchicken) [inline secret/store.go]: Doc is missing

File: secret/store.go [hunk 52]
```diff
@@ -0,0 +1,182 @@
...
+	log := logger.WithFields(logrus.Fields{"tag": "DeleteSecret"})
+	log.Debugf("Delete sectret: %s", fmt.Sprintf("secret/orgs/%s/%s", organizationID, secretID))
```
Comment (baluchicken) [inline secret/store.go]: typo in secret

Comment (baluchicken) [reply secret/store.go]: Do not use fmt.sprintf inside a logger which does the same thing.

File: api/organization.go [hunk 53]
```diff
@@ -0,0 +1,150 @@
...
+		log.Info(message + ": " + err.Error())
+		c.AbortWithStatusJSON(http.StatusInternalServerError, components.ErrorResponse{
+			Code:    http.StatusInternalServerError,
+			Message: message,
+			Error:   message,
+		})
+		return
+	}
+
+	if len(organizations) != 1 {
```
Comment (stoader) [inline api/organization.go]: Shouldn't we check for `len(organisations) == 0` ?

File: secret/store.go [hunk 54]
```diff
@@ -0,0 +1,182 @@
...
+	_, err := ss.logical.Delete(fmt.Sprintf("secret/orgs/%s/%s", organizationID, secretID))
+	return err
+}
+
+func (ss *SecretStore) Store(path string, value CreateSecretRequest) error {
```
Comment (baluchicken) [inline secret/store.go]: Doc is missing

File: secret/store.go [hunk 55]
```diff
@@ -0,0 +1,182 @@
...
+	log := logger.WithFields(logrus.Fields{"tag": "StoreSecret"})
+	log.Infof("Start storing secret")
+	data := map[string]interface{}{"value": value}
+	if _, err := ss.logical.Write(path, data); err != nil {
+		return errors.Wrap(err, "Error during storing secret")
+	}
+	return nil
+}
+
+func (ss *SecretStore) Get(organizationID string, secretID string) (*SecretsItemResponse, error) {
```
Comment (baluchicken) [inline secret/store.go]: Doc is missing

File: api/organization.go [hunk 56]
```diff
@@ -0,0 +1,150 @@
...
+		message := fmt.Sprintf("organization not found: %q", orgidParam)
```
Comment (stoader) [inline api/organization.go]: If we need to check for `len(organizations) > 1` as well than this error message is missleading.

File: secret/store.go [hunk 57]
```diff
@@ -0,0 +1,182 @@
...
+	secretPath := fmt.Sprintf("secret/orgs/%s/%s", organizationID, secretID)
```
Comment (baluchicken) [inline secret/store.go]: Do we need this variable?

File: secret/store.go [hunk 58]
```diff
@@ -0,0 +1,182 @@
...
+	parsedValues := make(map[string]string)
+	for k, v := range secretData["values"].(map[string]interface{}) {
+		parsedValues[k] = v.(string)
+	}
+	secretResp.Values = parsedValues
+
+	return secretResp, nil
+}
+
+func (ss *SecretStore) List(organizationID string) ([]SecretsItemResponse, error) {
```
Comment (baluchicken) [inline secret/store.go]: Doc is missing

File: api/organization.go [hunk 59]
```diff
@@ -0,0 +1,150 @@
...
+
+//GetOrganizations returns all organizations the user belongs to or a specific one from those by id
+func GetOrganizations(c *gin.Context) {
+	log := logger.WithFields(logrus.Fields{"tag": "GetOrganizations"})
+	log.Info("Fetching organizations")
+
+	user := auth.GetCurrentUser(c.Request)
+
+	idParam := c.Param("id")
+	id, err := strconv.ParseUint(idParam, 10, 32)
```
Comment (stoader) [inline api/organization.go]: This param seems to be optional thus we should try parse it only if specified.

File: api/organization.go [hunk 60]
```diff
@@ -0,0 +1,150 @@
...
+		log.Info(message + ": " + err.Error())
+		c.JSON(http.StatusInternalServerError, components.ErrorResponse{
+			Code:    http.StatusInternalServerError,
+			Message: message,
+			Error:   message,
+		})
+	} else if id == 0 {
+		c.JSON(http.StatusOK, organizations)
+	} else if len(organizations) != 0 {
+		c.JSON(http.StatusOK, organizations[0])
```
Comment (stoader) [inline api/organization.go]: If an `id` is passed in that there should be exactly one match. If more than one organisation is found than that is an error case.

File: cluster/aws.go [hunk 61]
```diff
@@ -446,6 +480,25 @@ func (c *AWSCluster) UpdateCluster(request *components.UpdateClusterRequest) err
...
+	}
+	clusterSecret, err := GetSecret(c)
+	if err != nil {
+		return err
+	}
+	if clusterSecret.SecretType != secret.Amazon {
+		return errors.Errorf("missmatch secret type %s versus %s", clusterSecret.SecretType, secret.Amazon)
+	}
+	awsCred := credentials.NewStaticCredentials(
+		clusterSecret.Values["AWS_ACCESS_KEY_ID"],
```
Comment (stoader) [inline cluster/aws.go]: `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` should be made a const as its used in multiple places.

File: secret/store.go [hunk 62]
```diff
@@ -0,0 +1,182 @@
...
+	return uuid.NewV4().String()
+}
+
+var DefaultRules = map[string][]string{
+	Amazon: {
+		"AWS_ACCESS_KEY_ID",
+		"AWS_SECRET_ACCESS_KEY",
+	},
+	Azure: {
+		"AZURE_CLIENT_ID",
```
Comment (stoader) [inline secret/store.go]: These literals should be factored out in a shared places as it used in multiple places

**Pull request merged**